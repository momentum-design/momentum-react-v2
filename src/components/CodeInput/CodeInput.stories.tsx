import React, { useState } from 'react';
import { Story } from '@storybook/react';

import CodeInput, { CodeInputProps } from './';
import { Message } from '../InputMessage/InputMessage.types';

const messageArrOptions = {
  empty: [],
  error: [{ message: 'Error message', level: 'error' }],
  success: [{ message: 'Success message', level: 'success' }],
  warning: [{ message: 'Warning message', level: 'warning' }],
};

export default {
  title: 'Momentum UI/CodeInput',
  component: CodeInput,
  argTypes: {
    className: {
      defaultValue: undefined,
      description:
        'If present, the class name will be added to the underlying component. Used to override styles by consumers.',
      control: { type: 'text' },
      table: {
        type: {
          summary: 'string',
        },
        defaultValue: {
          summary: undefined,
        },
      },
    },
    numDigits: {
      description: 'Number of boxes',
      control: { type: 'number' },
    },
    messageArr: {
      description: 'The list of messages to be passed in',
      options: Object.keys(messageArrOptions),
      mapping: messageArrOptions,
      control: { type: 'select' },
    },
    disabled: {
      description: 'Whether the input is disabled or not',
      control: { type: 'boolean' },
    },
    onComplete: {
      description: 'What happens when the code is complete',
      options: ['show error', 'go disabled', 'alert'],
      control: { type: 'select' },
    },
    onChange: {
      description: 'If present, the this handler will fire whenever the code changes',
      options: ['alert'],
      control: { type: 'select' },
    },
  },
};

interface StoryProps extends CodeInputProps {
  messageArr: [];
  onComplete;
  onChange;
}

const Template: Story<StoryProps> = (args) => {
  const { onComplete, onChange } = args;
  const [messageArrInt, setMessageArr] = useState<Message[]>(args.messageArr);
  const [isDisabled, setDisabled] = useState(args.disabled);

  return (
    <CodeInput
      numDigits={args.numDigits}
      messageArr={messageArrInt}
      disabled={isDisabled}
      onComplete={(code) => {
        switch (onComplete) {
          case 'show error':
            setMessageArr([{ message: 'test', level: 'error' }]);
            break;
          case 'go disabled':
            setDisabled(true);
            break;
          case 'alert':
            alert(`onComplete: code is ${code}`);
        }
      }}
      onChange={(code) => {
        if (onChange === 'alert') {
          alert(`onChange: code is ${code}`);
        }
      }}
    />
  );
};

const Example = Template;

Example.args = {
  numDigits: 6,
  messageArr: 'empty',
  disabled: false,
  onCompleteChoice: 'show error',
};

export { Example };
