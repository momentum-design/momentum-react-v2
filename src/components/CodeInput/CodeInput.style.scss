.md-code-input-wrapper {
  &:focus {
    box-shadow: none;
  }

  .md-code-input-container {
    height: 100%;
    width: 100%;

    &:focus {
      box-shadow: none;
    }

    .md-code-input-input {
      &:focus {
        box-shadow: none;
      }
    }

    .md-code-input-character {
      flex: auto;
      cursor: text;
      width: 2.75rem;
      height: 3rem;
      border: 1px solid var(--mds-color-theme-outline-input-normal);
      color: var(--mds-color-theme-text-primary-normal);
      background-color: var(--mds-color-theme-background-solid-primary-normal);
      box-sizing: border-box;
      border-radius: 0.5rem;
      font-size: 1.625rem;
      line-height: 3rem;

      &:not(:first-child) {
        margin-left: 0.5rem;
      }

      &:hover {
        border: 1px solid var(--mds-color-theme-outline-input-normal);
        color: var(--mds-color-theme-text-primary-normal);
        background-color: var(--mds-color-theme-background-primary-hover);
      }

      &:active {
        border-color: var(--mds-color-theme-outline-input-active);
        color: var(--mds-color-theme-text-primary-normal);
        background-color: var(--mds-color-theme-background-primary-active);
      }

      &--selected:empty {
        &:before {
          font-weight: 250;
          content: '|';
          animation: 1s blink step-end infinite;
        }

        border-color: var(--mds-color-theme-outline-theme-normal);
        color: var(--mds-color-theme-text-primary-normal);
        background-color: var(--mds-color-theme-background-primary-active);
      }
    }
  }

  @keyframes blink {
    from,
    to {
      color: transparent;
    }

    50% {
      color: var(--mds-color-theme-text-primary-normal);
    }
  }

  &[data-level='error'] {
    .md-code-input-container {
      .md-code-input-character {
        border-color: var(--mds-color-theme-outline-cancel-normal);
        color: var(--mds-color-theme-text-primary-normal);
        background-color: var(--mds-color-theme-background-alert-error-normal);

        &:hover {
          border-color: var(--mds-color-theme-outline-cancel-normal);
          color: var(--mds-color-theme-text-primary-normal);
          background-color: var(--mds-color-theme-background-primary-hover);
        }

        &:active {
          border-color: var(--mds-color-theme-outline-cancel-normal);
          color: var(--mds-color-theme-text-primary-normal);
          background-color: var(--mds-color-theme-background-primary-hover);
        }
      }
    }
  }

  input:disabled ~ .md-code-input-container {
    .md-code-input-character {
      cursor: default;
      border-color: var(--mds-color-theme-background-primary-ghost);
      color: var(--mds-color-theme-text-primary-normal);
      background-color: var(--mds-color-theme-background-primary-disabled);
    }
  }

  .md-code-input-message {
    margin-top: 1rem;
  }
}
