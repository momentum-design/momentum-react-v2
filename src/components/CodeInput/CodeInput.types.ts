import { Message } from '../InputMessage/InputMessage.types';

export interface Props {
  /**
   * Custom class to be able to override the component's CSS
   */
  className?: string;
  /**
   * numDigits: the number of input boxes in the code input
   */
  numDigits: number;
  /**
   * onChange: callback that fires whenever the entered code changes
   */
  onChange?: (code: string) => void;
  /**
   * onComplete: callback that fires when the user enters the final digit
   */
  onComplete?: (code: string) => void;
  /**
   * ariaLabel: the aria-label passed through to the input
   */
  ariaLabel?: string;
  /**
   * messageArr: array of messages and their types
   */
  messageArr?: Message[];
  /**
   * disabled: whether code input is disabled
   */
  disabled?: boolean;
  /**
   * visibleLabelId: the name field to be passed through to the input
   */
  visibleLabelId?: string;
}
