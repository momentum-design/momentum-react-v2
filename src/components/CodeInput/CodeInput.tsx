import React, { ReactElement, useState, useEffect, useRef } from 'react';
import VerificationInput from 'react-verification-input';

import './CodeInput.style.scss';
import { Props } from './CodeInput.types';
import InputMessage, { getFilteredMessages } from '../InputMessage';
import classnames from 'classnames';
import { STYLE } from './CodeInput.constants';

const CodeInput: React.FC<Props> = (props: Props): ReactElement => {
  const {
    numDigits,
    onChange = () => {
      /* Optional */
    },
    onComplete = () => {
      /* Optional */
    },
    ariaLabel,
    ariaDescribedby,
    messageArr = [],
    disabled = false,
    className,
    inputId,
  } = props;

  const [internalMessageArray, setInternalMessageArray] = useState(messageArr);
  const [isComplete, setComplete] = useState(false);
  const [value, setValue] = useState('');

  const [messageType, messages] = getFilteredMessages(internalMessageArray);

  const firstUpdate = useRef(true);
  useEffect(() => {
    if (firstUpdate.current) {
      firstUpdate.current = false;
      return;
    }
    onChange(value);
    if (value.length === numDigits) {
      onComplete(value);
      setComplete(true);
    }
  }, [value]);

  useEffect(() => {
    setInternalMessageArray(messageArr);
  }, [JSON.stringify(messageArr)]);

  return (
    <div className={classnames(STYLE.wrapper, className)} data-level={messageType}>
      <VerificationInput
        inputProps={{
          'aria-label': ariaLabel,
          'aria-describedby': ariaDescribedby,
          id: inputId,
          name: inputId,
          disabled,
        }}
        length={numDigits}
        onChange={setValue}
        onFocus={() => {
          if (isComplete) {
            // When completed, refocusing the input will clear the code and any messages
            setComplete(false);
            setValue('');
            setInternalMessageArray([]);
          }
        }}
        value={value}
        removeDefaultStyles={true}
        validChars="0-9"
        placeholder=""
        classNames={{
          container: STYLE.container,
          character: STYLE.character,
          characterInactive: STYLE.characterInactive,
          characterSelected: STYLE.characterSelected,
        }}
      />
      {messages && (
        <div className={STYLE.messages}>
          {messages.map((m, i) => (
            <InputMessage
              className={STYLE.message}
              message={m}
              key={`input-message-${i}`}
              level={messageType}
            />
          ))}
        </div>
      )}
    </div>
  );
};

/**
 * Numeric code input for use with confirmation codes
 */

export default CodeInput;
