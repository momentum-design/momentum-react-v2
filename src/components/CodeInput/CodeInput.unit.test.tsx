import CodeInput from '.';
import { mount } from 'enzyme';
import React, { useState } from 'react';
import { act } from 'react-dom/test-utils';

// TODO: swap for test util when available
const waitForAsync = (wrapper) =>
  new Promise<void>((resolve) =>
    setTimeout(() => {
      wrapper.update();

      return resolve();
    }, 0)
  );

describe('CodeInput', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      const container = mount(<CodeInput numDigits={6} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have correct number of boxes', () => {
      const component = mount(<CodeInput numDigits={6} />).childAt(0);
      expect(component.find('.md-code-input-container').length).toBe(1);
      expect(component.find('.md-code-input-character').length).toBe(6);
    });

    it('disabled code input should have input disabled', () => {
      const component = mount(<CodeInput numDigits={6} disabled />).childAt(0);
      expect(component.find('input').props().disabled).toBeTruthy();
    });

    it('input accepts prop aria-label', () => {
      const component = mount(<CodeInput numDigits={6} ariaLabel="enter 6 digit code" />).childAt(
        0
      );
      expect(component.find('input').prop('aria-label')).toBeDefined();
      expect(component.find('input').prop('aria-label')).toBe('enter 6 digit code');
    });

    it('input accepts prop name and id', () => {
      const component = mount(<CodeInput numDigits={6} inputId="test id" />).childAt(0);
      expect(component.find('input').prop('name')).toBeDefined();
      expect(component.find('input').prop('id')).toBeDefined();
      expect(component.find('input').prop('name')).toBe('test id');
      expect(component.find('input').prop('id')).toBe('test id');
    });

    it('when a message array is passing in messages should be displayed', async () => {
      let component;
      await act(async () => {
        component = mount(
          <CodeInput numDigits={6} messageArr={[{ message: 'test', level: 'error' }]} />
        );
      });
      await waitForAsync(component);
      const message = component.find('.md-input-message-message');
      expect(message.length).toBe(1);
      expect(message.props()['message-level']).toEqual('error');
    });

    it('should have custom class if provided', () => {
      const testClass = 'testClass';

      const wrapper = mount(<CodeInput numDigits={6} className={testClass} />);
      const element = wrapper.find(CodeInput).getDOMNode();

      expect(element.classList.contains(testClass)).toBe(true);
    });
  });

  describe('error on complete', () => {
    const ExampleContainer = () => {
      const [msgArr, setMsgArr] = useState([]);

      const onComplete = () => {
        setMsgArr([{ message: 'test', level: 'error' }]);
      };

      return <CodeInput onComplete={onComplete} numDigits={6} messageArr={msgArr} />;
    };

    it('displays a message when the message array changes', async () => {
      let component;

      await act(async () => {
        component = mount(<ExampleContainer />);
      });
      const message = component.find('.md-input-message-message');
      expect(message.length).toBe(0);

      let codeInput;
      await act(async () => {
        codeInput = component.find(CodeInput);
        codeInput.simulate('click');
        codeInput
          .find('input')
          .hostNodes()
          .simulate('change', { target: { value: '123456' } });
      });

      await waitForAsync(component);
      const message2 = component.find('.md-input-message-message');
      expect(message2.length).toEqual(1);
    });
  });

  describe('digit entry', () => {
    it('fires onChange when digits are entered', () => {
      const spy = jest.fn();
      const codeInput = mount(<CodeInput numDigits={3} onChange={spy} />);
      codeInput.simulate('click');

      const testInput = (value) => {
        codeInput.find('input').hostNodes().simulate('change', { target: { value } });
        expect(codeInput.find('input').props().value).toEqual(value);
        expect(spy).toBeCalledWith(value);
        spy.mockClear();
      };

      testInput('1');
      testInput('12');
      testInput('123');
      testInput('45');
      testInput('6');
    });

    it('fires onComplete when number of digits reached', () => {
      const spy = jest.fn();
      const codeInput = mount(<CodeInput numDigits={3} onComplete={spy} />);
      codeInput.simulate('click');

      const testInput = (value, expectOnComplete) => {
        codeInput.find('input').hostNodes().simulate('change', { target: { value } });
        expect(codeInput.find('input').props().value).toEqual(value);
        if (expectOnComplete) {
          expect(spy).toBeCalledWith(value);
        } else {
          expect(spy).not.toHaveBeenCalled();
        }
        spy.mockClear();
      };

      testInput('1', false);
      testInput('12', false);
      testInput('123', true);
      testInput('45', false);
      testInput('6', false);
      testInput('789', true);
    });
  });
});
