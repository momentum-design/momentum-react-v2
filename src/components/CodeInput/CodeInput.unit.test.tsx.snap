// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeInput snapshot should match snapshot 1`] = `
<CodeInput
  numDigits={6}
>
  <div
    className="md-code-input-wrapper"
    data-level="none"
  >
    <VerificationInput
      autoFocus={false}
      classNames={
        {
          "character": "md-code-input-character",
          "characterInactive": "md-code-input-character--inactive",
          "characterSelected": "md-code-input-character--selected",
        }
      }
      containerProps={
        {
          "className": "md-code-input-container",
        }
      }
      inputProps={
        {
          "aria-describedby": undefined,
          "aria-label": undefined,
          "className": "md-code-input-input",
          "disabled": false,
          "id": undefined,
          "name": undefined,
        }
      }
      length={6}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      validChars="0-9"
      value=""
    >
      <div
        className="vi__container md-code-input-container"
        data-testid="container"
        onClick={[Function]}
      >
        <input
          className="vi md-code-input-input"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onSelect={[Function]}
          spellCheck={false}
          value=""
        />
        <div
          className="vi__character md-code-input-character"
          data-testid="character-0"
          id="field-0"
          key="0"
          onClick={[Function]}
        />
        <div
          className="vi__character md-code-input-character vi__character--inactive md-code-input-character--inactive"
          data-testid="character-1"
          id="field-1"
          key="1"
          onClick={[Function]}
        />
        <div
          className="vi__character md-code-input-character vi__character--inactive md-code-input-character--inactive"
          data-testid="character-2"
          id="field-2"
          key="2"
          onClick={[Function]}
        />
        <div
          className="vi__character md-code-input-character vi__character--inactive md-code-input-character--inactive"
          data-testid="character-3"
          id="field-3"
          key="3"
          onClick={[Function]}
        />
        <div
          className="vi__character md-code-input-character vi__character--inactive md-code-input-character--inactive"
          data-testid="character-4"
          id="field-4"
          key="4"
          onClick={[Function]}
        />
        <div
          className="vi__character md-code-input-character vi__character--inactive md-code-input-character--inactive"
          data-testid="character-5"
          id="field-5"
          key="5"
          onClick={[Function]}
        />
      </div>
      <style
        dangerouslySetInnerHTML={
          {
            "__html": [
              [
                28,
                ".vi {
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  box-sizing: border-box;
  position: absolute;
  color: transparent;
  background: transparent;
  caret-color: transparent;
  outline: none;
  border: 0 none transparent;
}

.vi::-ms-reveal,
.vi::-ms-clear {
  display: none;
}

.vi::selection {
  background: transparent;
}

/* :where() gives the styles specificity 0, which makes them overridable */
:where(.vi__container) {
  position: relative;
  display: flex;
  gap: 8px;
  height: 50px;
  width: 300px;
}

:where(.vi__character) {
  height: 100%;
  flex-grow: 1;
  flex-basis: 0;
  text-align: center;
  font-size: 36px;
  line-height: 50px;
  color: black;
  background-color: white;
  border: 1px solid black;
  cursor: default;
  user-select: none;
  box-sizing: border-box;
}

:where(.vi__character--inactive) {
  color: dimgray;
  background-color: lightgray;
}

:where(.vi__character--selected) {
  outline: 2px solid cornflowerblue;
  color: cornflowerblue;
}
",
                "",
              ],
            ],
          }
        }
      />
    </VerificationInput>
    <div
      className="md-code-input__messages"
    />
  </div>
</CodeInput>
`;
